opt: Add index support to the optbase.Catalog interface

andy-kimball opened the pull request:
Add support for primary and secondary indexes to the optbase.Catalog
interface. This consists of additional methods on optbase.Table to
get the indexes, as well as an additional Index interface and
IndexColumn struct. Provide an implementation of the expanded interface
over SchemaResolver. This requires adding wrapper objects for Table
and Index descriptors, in order to link index columns back to table
columns, as well as to add a ColumnID lookup map.

Also rename and split executor_opt_interface.go and provide a catalog
table formatting method for debugging and testing.

Release note: None

andy-kimball committed:
opt: Add index support to the optbase.Catalog interface

Add support for primary and secondary indexes to the optbase.Catalog
interface. This consists of additional methods on optbase.Table to
get the indexes, as well as an additional Index interface and
IndexColumn struct. Provide an implementation of the expanded interface
over SchemaResolver. This requires adding wrapper objects for Table
and Index descriptors, in order to link index columns back to table
columns, as well as to add a ColumnID lookup map.

Also rename and split executor_opt_interface.go and provide a catalog
table formatting method for debugging and testing.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23000)
<!-- Reviewable:end -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/optbase/catalog.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L62a7i5Mjzr6gJ8-o52:-L62a7i5Mjzr6gJ8-o53:b-jmzg7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa06d8adcdcc2e049908180c073f0c7d61546d18/pkg/sql/optbase/catalog.go#L60)):*
> ```Go
> 
> 	// Ordinal is the ordinal position of the indexed column in the table being
> 	// indexed. It can be passed to the Table.Column method to retrieve a table
> ```

The second sentence is not very useful, given that we already have a Column above. Maybe just mention the value is between 0 and Table.ColumnCount().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-:-L62aPgi_2mG2p35Yn42:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/optbase/catalog.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L62a7i5Mjzr6gJ8-o52:-L62x2ukU42lCBSDaaC7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa06d8adcdcc2e049908180c073f0c7d61546d18/pkg/sql/optbase/catalog.go#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The second sentence is not very useful, given that we already have a Column above. Maybe just mention the value is between 0 and Table.ColumnCount().
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 12 of 13 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_catalog.go, line 95 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63K785zyRPcKeMgSFo:-L63K785zyRPcKeMgSFp:b-ilotgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L95)):*
> ```Go
> }
> 
> // Name is part of the optbase.Table interface.
> ```

`Name` -> `TabName`

---

*[pkg/sql/opt_catalog.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63LSVJMQzqsm7KfIwp:-L63LSVJMQzqsm7KfIwq:b-tiu0mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L126)):*
> ```Go
> 	// Check to see if there's already a wrapper for this index descriptor.
> 	if ot.wrappers == nil {
> 		ot.wrappers = make(map[*sqlbase.IndexDescriptor]*optIndex)
> ```

You can allocate the map with capacity `len(ot.desc.Indexes)`

---

*[pkg/sql/opt_catalog.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63L6e_O2T8CJnBDd1K:-L63L6e_O2T8CJnBDd1L:bjd0lq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L140)):*
> ```Go
> func (ot *optTable) lookupColumnOrdinal(colID sqlbase.ColumnID) int {
> 	if ot.colMap == nil {
> 		ot.colMap = make(map[sqlbase.ColumnID]int)
> ```

You can allocate the map with capacity `len(ot.desc.Columns)`

---

*[pkg/sql/opt_catalog.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63X9iaNodREGUCvQUh:-L63WkyH6nA5r6ZSUIUJ:b8bz190) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L208)):*
> ```Go
> 	length = len(oi.desc.ExtraColumnIDs)
> 	if i < length {
> 		// Extra columns are in the key only if this is not a unique index.
> ```

I think this comment is more confusing than helpful - why do we care about what is in the key or not in the key?  The distinction doesn't seem relevant here.

---

*[pkg/sql/opt_catalog.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63X3SuIZ5qKIKOdwo9:-L63X3SuIZ5qKIKOdwoA:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L213)):*
> ```Go
> 	}
> 
> 	// Storing columns are never in the key.
> ```

Ditto

---

*[pkg/sql/opt/exec/execbuilder/testdata/catalog, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63V8RPS-T9nUXYYUrG:-L63V8RPS-T9nUXYYUrH:b8v13f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt/exec/execbuilder/testdata/catalog#L53)):*
> ```
>     PRIMARY KEY (x, y DESC),
>     UNIQUE INDEX myindex (x DESC, z) STORING (i),
>     INDEX (x) STORING(i)
> ```

What about a test for
`INDEX (z)`? In that case would `x` and `y` be added after `z`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-:-L63YgPWV2U0LbS5p-2h:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_catalog.go, line 95 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63K785zyRPcKeMgSFo:-L64TB3KbFdrMmcjmf6Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L95)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`Name` -> `TabName`
</blockquote></details>

Done.

---

*[pkg/sql/opt_catalog.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63LSVJMQzqsm7KfIwp:-L64TMka15V8Rpbb5aQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L126)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You can allocate the map with capacity `len(ot.desc.Indexes)`
</blockquote></details>

Done.

---

*[pkg/sql/opt_catalog.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63L6e_O2T8CJnBDd1K:-L64TQhAb4MJpwr5ql1u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L140)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You can allocate the map with capacity `len(ot.desc.Columns)`
</blockquote></details>

Done.

---

*[pkg/sql/opt_catalog.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63X9iaNodREGUCvQUh:-L64TbS6XbcMi5yY6Wc_:b-u6jxjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L208)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think this comment is more confusing than helpful - why do we care about what is in the key or not in the key?  The distinction doesn't seem relevant here.
</blockquote></details>

This was holdover from older version where it made more sense. Removed.

---

*[pkg/sql/opt_catalog.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63X3SuIZ5qKIKOdwo9:-L64Tdl4Az83FGcNR8Go:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt_catalog.go#L213)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/sql/opt/exec/execbuilder/testdata/catalog, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000#-L63V8RPS-T9nUXYYUrG:-L64U7S7mjI611yhHKqn:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/708998f12643a032dcd813cb8ea36413a610db11/pkg/sql/opt/exec/execbuilder/testdata/catalog#L53)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What about a test for
`INDEX (z)`? In that case would `x` and `y` be added after `z`?
</blockquote></details>

Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23000)*
<!-- Sent from Reviewable.io -->
